plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+		} else {
+			padded_len = 64;
+			nbytes = padded_len - unaligned_bytes;
+		}
+
+		/* set nvm read offset in dwords */
+		fu_dell_dock_usb4_write_u32to_buf (NVM_OFFSET_TO_METADATA(nvm_addr),
+						   metadata);
+
+		/* and length field in dwords, note 0 means 16 dwords */
+		metadata[3] = (padded_len / 4) & 0xf;
+
+		/* ask hub to read up to 64 bytes from NVM to mbox data regs */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_READ,
+						      metadata, error)) {
+			g_prefix_error (error, "hub NVM read error:");
+			return FALSE;
+		}
+		/* read the data from mbox data regs into our buffer */
+		if (!fu_dell_dock_usb4_mbox_data_read (device, tmpbuf,
+						       padded_len, error)) {
+			g_prefix_error (error, "hub firmware mbox data read error:");
+			return FALSE;
+		}
+		if (!fu_memcpy_safe (buf, length, 0x0,
+				     tmpbuf, sizeof(tmpbuf), unaligned_bytes,
+				     nbytes, error))
+			return FALSE;
+
+		buf += nbytes;
+		nvm_addr += nbytes;
+		length -= nbytes;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_nvm_write (FuDevice *device, const guint8 *buf,
+				 guint32 length, guint32 nvm_addr,
+				 GError **error)
+{
+	guint8 metadata[4];
+	guint32 bytes_done = 0;
+	guint32 bytes_total = length;
+
+	if (nvm_addr % 4) {
+		g_prefix_error (error,
+			       "Invalid NVM write offset 0x%x, must be DW aligned:",
+			        nvm_addr);
+		return FALSE;
+	}
+	if (length < 64 || length % 64) {
+		g_prefix_error (error,
+			        "Invalid NVM length 0x%x, must be 64 byte aligned:",
+			        length);
+		return FALSE;
+	}
+
+	/* 1. Set initial offset, must be DW aligned */
+	fu_dell_dock_usb4_write_u32to_buf (NVM_OFFSET_TO_METADATA(nvm_addr),
+					   metadata);
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_SET_OFFSET,
+					      metadata, error)) {
+		g_prefix_error (error, "hub NVM set offset error:");
+		return FALSE;
+	}
+
+	/* 2 Write data in 64 byte blocks */
+	fu_device_set_progress_full (device, bytes_done, bytes_total);
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	while (length > 0)  {
+		/* write data to mbox data regs */
+		if (!fu_dell_dock_usb4_mbox_data_write (device, buf, 64,
+							error)) {
+			g_prefix_error (error,
+					"hub mbox data write error:");
+			return FALSE;
+		}
+		/* ask hub to write 64 bytes from data regs to NVM */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_WRITE,
+						      NULL, error)) {
+			g_prefix_error (error, "hub NVM write operation error:");
+			return FALSE;
+		}
+		buf += 64;
+		length -= 64;
+		fu_device_set_progress_full (device, bytes_done += 64,
+					     bytes_total);
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	return TRUE;
+}
+
+/* Issue a nvm auth write operation */
+static gboolean
+fu_dell_dock_usb4_hub_nvm_authenticate (FuDevice *device, GError **error)
+{
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_AUTH_WRITE,
+					      NULL, error)) {
+		g_print ("NVM authenticate failed");
```

Comment:
keep in mind the client doesn't show this - the daemon does.  It's better to do g_prefix_error here so it flows back into the client.